File: python/second/rotate_image.py
class Solution(object):
    # O(n^2) time | O(1) space
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        l = 0
        r = len(matrix) - 1

        while l < r:
            top = l
            bottom = r

            for i in range(r - l):
                topLeft = matrix[top][l + i]
                matrix[top][l + i] = matrix[bottom - i][l]
                matrix[bottom - i][l] = matrix[bottom][r - i]
                matrix[bottom][r - i] = matrix[top + i][r]
                matrix[top + i][r] = topLeft
            l += 1
            r -= 1

matrix = [[1,2,3],[4,5,6],[7,8,9]]

solution = Solution()
solution.rotate(matrix)
print(matrix)

File: src/main/java/october_2025/RotateImage.java
package october_2025;

public class RotateImage {

    public static void main(String[] args) {
        int[][] matrix = {
                {1, 2, 3},
                {4, 5, 6},
                {7, 8, 9}
        };

        RotateImage rotateImage = new RotateImage();
        rotateImage.rotate(matrix);
        System.out.println();
    }

    // O(n^2) time | O(1) space
    public void rotate(int[][] matrix) {
        int l = 0;
        int r = matrix.length - 1;

        // l, top ++
        // r, bottom --

        while (l < r) {
            int top = l;
            int bottom = r;
            for (int i = 0; i < r - l; i++) {
                int topLeft = matrix[top + i][l];
                // top left = bottom left
                matrix[top + i][l] = matrix[bottom][l + i]; // 1 -> 7

                // bottom left = bottom right
                matrix[bottom][l + i] = matrix[bottom - i][r]; // 7 -> 9

                // bottom right = top right
                matrix[bottom - i][r] = matrix[top][r - i]; // 9 -> 3

                // top right = top left
                matrix[top][r - i] = topLeft; // 3 -> 1
            }
            l++;
            r--;
        }
    }

}
